### PR TITLE
Completed DCGAN tutorial

### PR DESCRIPTION
**Pull request by margaretmz**
I have completed the DCGAN tutorial with detailed explanations, a diagram and learning resources. Please review and approve the pull request.

### CODE REVIEW

**Comment by alextp** [comment]
I somehow can't get github's notebook visualizer to show me this file. It seems to be a single line of code as opposed to the many lines before your edit; do you know if there's a way to reexport the ipynb?

**Comment by margaretmz** [comment]
@alextp, I was able to show the notebook through the github notebook visualizer although at times it doesn't seem to work. I have reexported the notebook to better reflect the code diff. Looking forward to your review and feedback.

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 1]
```diff
@@ -1,78 +1,186 @@
...
+      ]
+    },
+    {
+      "metadata": {
+        "id": "ITZuApL56Mny",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "This tutorial demonstrates how to generate images of handwritten digits with **Deep Convolutional Generative Adverserial Networks** ([DCGAN](https://arxiv.org/pdf/1511.06434.pdf)). The code is written in [tf.keras](https://www.tensorflow.org/programmers_guide/keras) with [eager execution](https://www.tensorflow.org/programmers_guide/eager) enabled. "
```
Comment (alextp) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: adverserial -> adversarial

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 2]
```diff
@@ -1,78 +1,186 @@
...
-        "colab_type": "text",
-        "id": "ITZuApL56Mny"
+        "id": "2MbKJY38Puy9",
+        "colab_type": "text"
       },
+      "cell_type": "markdown",
       "source": [
-        "This notebook demonstrates how to generate images of handwritten digits using [tf.keras](https://www.tensorflow.org/programmers_guide/keras) and [eager execution](https://www.tensorflow.org/programmers_guide/eager). To do so, we use Deep Convolutional Generative Adverserial Networks ([DCGAN](https://arxiv.org/pdf/1511.06434.pdf)).\n",
+        "## What are GANs?\n",
+        "GANs standards for **Generative Adversarial Networks** and they are a type of deep **generative** models. MIT [Intro to Deep Learning](http://introtodeeplearning.com/) lecture on **Deep Generative Models** has a great intro to generative models as well as GANs. ([video](https://youtu.be/JVb54xhEw6Y) | [slides](http://introtodeeplearning.com/materials/2018_6S191_Lecture4.pdf)). We have included more learning resources on these subjects in the \"Learn more about GANs\" section at the end of the tutorial.\n",
```
Comment (alextp) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: "standards for" -> "stands for"

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 3]
```diff
@@ -1,78 +1,186 @@
...
         "\n",
-        "This model takes about ~30 seconds per epoch (using tf.contrib.eager.defun to create graph functions) to train on a single Tesla K80 on Colab, as of July 2018.\n",
+        " **Deep Convolutional GAN** (DCGAN) is a type of GANs and in this tutorial we will use DCGAN to generate MNIST digits.\n",
         "\n",
-        "Below is the output generated after training the generator and discriminator models for 150 epochs.\n",
+        "GANs can be used to generate new images that no one has seen before. The generator will generate fake images while the discriminator will classify whether the generated images are fake. An** equilibrium** will reach in the game when the generator makes data that looks identical to the training data and the discriminator can no longer tell the difference between the fake images (generated by the generator) and the real images (the training data). \n",
```
Comment (alextp) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: "an equilibrium will reach in the game" -> "the game will reach an equilibrium"

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 4]
```diff
@@ -1,78 +1,186 @@
...
+        "\n",
+        "![alt text](https://github.com/margaretmz/tensorflow/blob/margaret-dcgan/tensorflow/contrib/eager/python/examples/generative_examples/gans_diagram.png?raw=1)\n",
+        "\n",
+        "While the generator and discriminator competes against each other, the discriminator also teaches the generator . Over time the generator starts to produce images that resemble the training data that is fed into the discriminator, in this case the MNIST digits. Below is the output with images generated after training the generator and discriminator models for 150 epochs.\n",
```
Comment (alextp) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: "competes against" -> "compete against"

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 5]
```diff
@@ -327,96 +517,100 @@
...
       ]
     },
     {
+      "metadata": {
+        "id": "5QC5BABamh_c",
+        "colab_type": "text"
+      },
       "cell_type": "markdown",
+      "source": [
+        "Now it's time to put together the generator and discriminator to set up the Generative Adversarial Networks, as you wee the diagam in the beginning of the tutorial."
```
Comment (alextp) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: I think you did not mean to say "as you wee the diagram". See?

**Review by alextp — CHANGES_REQUESTED** [review]
Looks much improved; thanks! I only have a few minor spelling nits.

**Comment by margaretmz** [comment]
Thanks @alextp for your review and feedback! I made updates accordingly.

**Review by alextp — DISMISSED** [review]

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 6]
```diff
@@ -84,114 +192,135 @@
...
-        "colab": {},
+        "id": "VGLbvBEmjK0a",
         "colab_type": "code",
-        "id": "VGLbvBEmjK0a"
+        "colab": {}
       },
-      "outputs": [],
+      "cell_type": "code",
       "source": [
         "class Generator(tf.keras.Model):\n",
```
Comment (asimshankar) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Both this and the discriminator class seems like they will be well served by `tf.keras.Sequential`, making the code even more compact. Should we write them instead as:

```python
def make_generator_model():
  return tf.keras.Sequential([
    tf.keras.layers.Dense(7*7*64, use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Relu(),
    ...])
```

**Review by asimshankar — CHANGES_REQUESTED** [review]

**Comment by random-forests** [comment]
Thank you @margaretmz! I added some edits to the text (it was easier to commit the source of my Jupyter Notebook rather than copy / pasting them in, I hope I didn't clobber anything!)

**Comment by margaretmz** [comment]
Thank you @asimshankar for suggesting we use the Sequential model instead of model subclassing for the Generator and Discriminator classes. The current code is using  tf.contrib.eager.defun to boost performance. Can we still use defun with the Sequential model? My understanding is that defun is typically applied in the forward computation, backward computation for the gradients and applications of gradients to variables.

**Comment by margaretmz** [comment]
Thanks @random-forests for your edits to make the tutorial more concise! I verified that the edited notebook looks good.

**Comment by alextp** [comment]
@margaretmz defun and sequential should work well together, so this should be fine.

**Comment by random-forests** [comment]
@asimshankar, thanks for the suggestion on sequential. I added a note about that in the notebook. In this case, I'd prefer to keep subclassing (even though its more verbose), just so we have more examples on tf.org in this style - in case it'll help folks modify this for other projects down the road.

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 7]
```diff
@@ -1,694 +1,928 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {},
+    "colab_type": "code",
+    "id": "YfIk2es3hJEd"
+   },
+   "outputs": [],
+   "source": [
+    "from __future__ import absolute_import, division, print_function\n",
```
Comment (alextp) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Can we ditch the future imports? I think they're sad.

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 8]
```diff
@@ -1,694 +1,928 @@
...
+    "        disc_loss = discriminator_loss(real_output, generated_output)\n",
+    "        \n",
+    "      gradients_of_generator = gen_tape.gradient(gen_loss, generator.variables)\n",
+    "      gradients_of_discriminator = disc_tape.gradient(disc_loss, discriminator.variables)\n",
+    "      \n",
+    "      generator_optimizer.apply_gradients(zip(gradients_of_generator, generator.variables))\n",
+    "      discriminator_optimizer.apply_gradients(zip(gradients_of_discriminator, discriminator.variables))\n",
+    "\n",
+    "      \n",
+    "    if epoch % 1 == 0:\n",
```
Comment (alextp) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: epoch % 1 == 0 is a tautology

**Review by alextp — CHANGES_REQUESTED** [review]

**Comment by random-forests** [comment]
Silly CLA bot.

**Comment by asimshankar** [comment]
@margaretmz : Yup, it should work, see https://github.com/tensorflow/tensorflow/commit/96c7397e08a8fde1bdf069b4e7dd74b5a8d79a31 :)

@random-forests : I have a slight preference for using `Sequential` here since I think we do want to encourage that style, not only because of brevity, but also because it provides a bunch of other benefits in terms of serialization, introspection (e.g., iterating over the layer structure of a network) etc. Till a couple of months ago there were a bunch of pain points with `Sequential` in eager that have since been addressed. Long story short: I think using `Sequential` here will be good for users and instructive. But if you still feel strongly, I won't object.

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb (hunk 7 above)
Comment (margaretmz) [reply tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Removed the future imports.

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb (hunk 8 above)
Comment (margaretmz) [reply tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Removed.

**Comment by margaretmz** [comment]
I've made updates for your requested changes, @alextp.

Thanks @alextp and @asimshankar for confirming that Sequential and defun work well together. I agree with @random-forests - what we have right now in the tutorial will let the readers see both styles and there are plenty of examples on Sequential already.

**Review by alextp — DISMISSED** [review]

**Comment by margaretmz** [comment]
@asimshankar, @alextp and @random-forests - I updated the generator and discriminator models to use Sequential, and updated the text accordingly. I modified the generator model during the conversion and the network now generates resembling digits after only 50 epochs. Please review and provide your feedback. Thanks.

**Review by alextp — DISMISSED** [review]

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 9]
```diff
@@ -1,324 +1,428 @@
...
-        "    self.conv2 = tf.keras.layers.Conv2DTranspose(32, (5, 5), strides=(2, 2), padding='same', use_bias=False)\n",
-        "    self.batchnorm3 = tf.keras.layers.BatchNormalization()\n",
+        "def make_generator_model():\n",
+        "  return tf.keras.Sequential([\n",
+        "    tf.keras.layers.Dense(7*7*256, use_bias=False),\n",
+        "    tf.keras.layers.BatchNormalization(),\n",
+        "    tf.keras.layers.LeakyReLU(),\n",
+        "      \n",
+        "    tf.keras.layers.Reshape((7, 7, 256)),\n",
+        "    # Reshape layer to 7x7x64 \n",
```
Comment (asimshankar) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Optional: Instead of these comments, I wonder if it would make sense to use the Keras model's `output_shape` property and assertions.  That way the comment and code are always in sync. Though, you'll need to provide the `input_shape` argument to the first layer.

Something like:

```python
def make_generator_model():
  m = tf.keras.Sequential()
  m.add(tf.keras.layers.Dense(7*7*256, use_bias=False, input_shape=(100,))
  m.add(tf.keras.layers.BatchNormalization())
  m.add(tf.keras.layers.LeakyReLU())
  m.add(tf.keras.layers.Conv2DTranspose(128, (5, 5), strides=(1, 1), padding='same', use_bias=False))
  assert m.output_shape == (None, 7, 7, 64)
  m.add(...)
  ...
  assert m.output_shape == (None, 28, 28, 1)
  return m
```
(Actually, in this process, I noticed that the comments are a bit off :))

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 10]
```diff
@@ -1,324 +1,428 @@
...
+        "colab_type": "text",
+        "id": "6TSZgwc2BUQ-"
       },
-      "outputs": [],
+      "cell_type": "markdown",
       "source": [
-        "generator = Generator()\n",
-        "discriminator = Discriminator()"
+        "\n",
+        "This model takes about ~30 seconds per epoch to train on a single Tesla K80 on Colab, as of July 2018. Eager execution can sometimes be slower than executing the equivalent graph due to overheads of interpreting Python code. By using [tf.contrib.eager.defun](https://www.tensorflow.org/api_docs/python/tf/contrib/eager/defun) to create graph functions, we get a ~10 secs/epoch performance boost. This way we get the best of both eager execution (easier for debugging) and graph mode (better performance)."
```
Comment (asimshankar) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Optional: Instead of `defun`ing the model calls, did you try `defun`-ing the whole training step? That might yield some good improvements too (and will be simpler)?

So instead of:

```python
def train(dataset, epochs, noise_dim):
  for epoch in range(epochs):
    for images in dataset:
      ...
```

You could have:

```python
def train_step(images):
   noise = tf.random_normal(...)
   with tf.GradientTape() as tape:
      ...
   

def train(dataset, epochs):
  for epoch in epochs:
    for images in dataset:
      train_step(images)
    # All that display stuff
```

Then for faster training you could do:
```python
train_step = tf.contrib.eager.defun(train_step)
```

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 11]
```diff
@@ -327,142 +431,134 @@
...
-      "execution_count": 0,
       "metadata": {
-        "colab": {},
         "colab_type": "code",
-        "id": "2M7LmLtGEMQJ"
+        "id": "2M7LmLtGEMQJ",
+        "colab": {}
       },
-      "outputs": [],
+      "cell_type": "code",
```
Comment (asimshankar) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Why pass `noise_dim` as an argument to `train()` given that it is defined as almost a constant above?
The module level `noise_dim` constant and the function argument can be confusing. How about just having a `NOISE_DIM` constant and using that instead of creating a constant and then providing that as an argument of the same name to `train()`?

**Review by asimshankar — DISMISSED** [review]

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb (hunk 9 above)
Comment (margaretmz) [reply tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: @asimshankar, thanks for your suggestion! I made changes as you suggested.
It's indeed better than comments - yesterday I forgot to update the comment after I change the code.

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb (hunk 10 above)
Comment (margaretmz) [reply tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Updated to defun on train_steps instead of model calls. Performance boost is better at ~20 secs / epoch instead of ~10 secs / epoch (defun on model calls). Thanks!

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb (hunk 11 above)
Comment (margaretmz) [reply tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Good catch. Fixed.

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 12]
```diff
@@ -1,324 +1,405 @@
...
-        "    \n",
-        "    self.conv2 = tf.keras.layers.Conv2DTranspose(32, (5, 5), strides=(2, 2), padding='same', use_bias=False)\n",
-        "    self.batchnorm3 = tf.keras.layers.BatchNormalization()\n",
+        "def make_generator_model():\n",
+        "    model = tf.keras.Sequential()\n",
+        "    model.add(tf.keras.layers.Dense(7*7*256, use_bias=False, input_shape=(100,)))\n",
+        "    model.add(tf.keras.layers.BatchNormalization())\n",
+        "    model.add(tf.keras.layers.LeakyReLU())\n",
+        "      \n",
+        "    model.add(tf.keras.layers.Reshape((7, 7, 256)))\n",
```
Comment (asimshankar) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Optional: Maybe a comment here mentioning that `None` is the batch dimension?

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 13]
```diff
@@ -1,324 +1,405 @@
...
     {
+      "metadata": {
+        "id": "dw2tPLmk2pEP",
+        "colab_type": "code",
+        "colab": {}
+      },
       "cell_type": "code",
+      "source": [
+        "def make_discriminator_model():\n",
+        "    model = tf.keras.Sequential()\n",
```
Comment (asimshankar) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Optional: Since there are no asserts here, the original version (where the list was provided to the constructor) works just as well. Your call either way.

**Review by asimshankar — DISMISSED** [review]
Looks good to me, thanks!

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb [hunk 14]
```diff
@@ -327,170 +408,200 @@
...
+      "cell_type": "markdown",
+      "source": [
         "\n",
-        "      \n",
-        "    if epoch % 1 == 0:\n",
-        "      display.clear_output(wait=True)\n",
-        "      generate_and_save_images(generator,\n",
+        "This model takes about ~30 seconds per epoch to train on a single Tesla K80 on Colab, as of October 2018. \n",
+        "\n",
+        "Eager execution can sometimes be slower than executing the equivalent graph due to overheads of interpreting Python code. By using [tf.contrib.eager.defun](https://www.tensorflow.org/api_docs/python/tf/contrib/eager/defun) to create graph functions, we get a ~20 secs/epoch performance boost (from ~50 secs/epoch down to ~30 secs/epoch). This way we get the best of both eager execution (easier for debugging) and graph mode (better performance)."
```
Comment (asimshankar) [inline tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: I'd rephrase this a bit (so as to not blame just the Python interpreter). There may be other graph optimizations we're benefiting from here as well (constant folding, common subexpression elimination, layout changes etc.). Though, in this particular case it is probably the Python overhead (since we haven't enabled grappler passes on function yet). How about:

"Eager execution can be slower than executing the equivalent graph as it can't benefit from whole-program optimizations on the graph, and also incurs overheads of interpreting Python code. By using ..."

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb (hunk 12 above)
Comment (margaretmz) [reply tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Done

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb (hunk 13 above)
Comment (margaretmz) [reply tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: I left it as is because it's easier for the reader to experiment with assert, print or model.summary() etc. if they want to.

File: tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb (hunk 14 above)
Comment (margaretmz) [reply tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb]: Updated as suggested.

**Comment by margaretmz** [comment]
@asimshankar, @alextp and @random-forests, I have made updates per all your review feedback. Looking forward to your approval and PR merge. Thanks.

**Review by alextp — APPROVED** [review]

**Review by asimshankar — APPROVED** [review]
